org.eclipse.scout.rt.shared/src/main/java/org/eclipse/scout/rt/shared/extension/AbstractContributionComposite.java

```diff
@@ -58,63 +57,52 @@ private <T> void initContributionsMap(Object o, IInternalExtensionRegistry exten
-    }
     if (CollectionUtility.hasElements(contributionsForMe)) {
+      if (m_contributionsByClass == null) {
+        m_contributionsByClass = new HashMap<>(contributionsForMe.size(), 1.0f); // use fixed size and loadFactor
+      }
       for (Object contribution : contributionsForMe) {
         m_contributionsByClass.put(contribution.getClass(), contribution);
       }
     }
   }
 
   public <T> void resetContributionsByClass(Object o, Class<T> type) {
-    m_contributionsByType.remove(type);
-    m_contributionsByClass.values().removeIf(obj -> type.isAssignableFrom(obj.getClass()));
+    Assertions.assertNotNull(type, "Contribution type class must be specified");
+    if (m_contributionsByClass != null) {
+      m_contributionsByClass.values().removeIf(obj -> type.isAssignableFrom(obj.getClass()));
+    }
     IInternalExtensionRegistry extensionRegistry = BEANS.get(IInternalExtensionRegistry.class);
     initContributionsMap(o, extensionRegistry, type);
   }
 
-  private void readObject(ObjectInputStream s) throws IOException, ClassNotFoundException {
-    s.defaultReadObject();
-    if (m_contributionsByClass == null || m_contributionsByType == null) {
-      // ensure that the contributions have been initialized
-      IInternalExtensionRegistry extensionRegistry = BEANS.get(IInternalExtensionRegistry.class);
-      initContributionsMap(null, extensionRegistry, null);
+  protected Collection<Object> getAllContributionsInternal() {
+    if (m_contributionsByClass == null) {
+      return null;
     }
-  }
-
-  protected Collection<Object> geAllContributionsInternal() {
     return m_contributionsByClass.values();
   }
 
   @Override
   public List<Object> getAllContributions() {
-    return CollectionUtility.arrayList(geAllContributionsInternal());
+    return CollectionUtility.arrayList(getAllContributionsInternal());
   }
 
   @Override
-  @SuppressWarnings("unchecked")
   public <T> List<T> getContributionsByClass(Class<T> type) {
     if (type == null) {
       throw new IllegalArgumentException("Contribution type class must be specified.");
     }
 
-    ArrayList<T> contributionsOfType = (ArrayList<T>) m_contributionsByType.get(type);
-    if (contributionsOfType == null) {
-      Collection<Object> values = m_contributionsByClass.values();
-      contributionsOfType = new ArrayList<>(values.size());
-      for (Object o : values) {
-        if (type.isAssignableFrom(o.getClass())) {
-          T contribution = type.cast(o);
-          contributionsOfType.add(contribution);
-        }
-      }
-      contributionsOfType.trimToSize();
-      m_contributionsByType.put(type, contributionsOfType);
+    if (m_contributionsByClass == null) {
+      return new ArrayList<>(0);
     }
-    return CollectionUtility.arrayList(contributionsOfType);
+
+    return m_contributionsByClass.values()
+        .stream()
+        .filter(o -> type.isAssignableFrom(o.getClass()))
```

Comment:
```suggestion
        .filter(o -> type.isInstance(o))
```

or as method reference
```suggestion
        .filter(type::isInstance)
```